### PR TITLE
Allow "aws cloudformation package" to succeed when a Resource is not a key-value pair

### PR DESCRIPTION
**Pull request by arthurboghossian**
*Issue #, if available:* #8075 

*Description of changes:*

Allow "aws cloudformation package" to succeed when a Resource is not a key-value pair.

This includes supporting the new `Fn::ForEach` intrinsic function, which can be provided as a key-value pair within the Resources section, where the value is not a key-value pair, as expected, but a list.

Note: was able to successfully run `aws cloudformation package` and `aws cloudformation deploy` on CloudFormation templates that contained the `Fn::ForEach` intrinsic function.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by arthurboghossian** [comment]
Added a warning message that local artifacts aren't currently supported in Resources that aren't defined as key-value pairs

**Comment by arthurboghossian** [comment]
In the latest commit, modified artifact_explorer, where instead of skipping over resources that aren't dictionaries, modifying the export method to be able to handle the Fn::ForEach intrinsic function by recursively calling the export function on the fragment property of Fn::ForEach

**Review by kyleknap — COMMENTED** [review]
Looks good. I just had some smaller comments.

File: awscli/customizations/cloudformation/artifact_exporter.py [hunk 1]
```diff
@@ -659,9 +659,23 @@ def export(self):
 
         self.template_dict = self.export_global_artifacts(self.template_dict)
 
-        for resource_id, resource in self.template_dict["Resources"].items():
```
[the remark this answers is not in the document]
Comment (arthurboghossian) [reply awscli/customizations/cloudformation/artifact_exporter.py:662]: I've ran the command and added the generated changelog

File: awscli/customizations/cloudformation/artifact_exporter.py [hunk 2]
```diff
@@ -659,9 +659,23 @@ def export(self):
...
+
+        return self.template_dict
+
+    def export_resources(self, resource_dict):
+        for resource_id, resource in resource_dict.items():
+
+            if resource_id.startswith("Fn::ForEach"):
+                if not isinstance(resource, list) or len(resource) < 3:
+                    raise ValueError("Fn::ForEach with name {0} has invalid format".format(resource_id))
+                if isinstance(resource[2], dict):
```
[the remark this answers is not in the document]
Comment (arthurboghossian) [reply awscli/customizations/cloudformation/artifact_exporter.py:672]: There are future enhancements to Fn::ForEach that would result in `resource[2]` not being a dictionary, but if that's the case, we shouldn't throw an exception, and shouldn't export resources as there wouldn't be any within `resource[2]`. So the only time there will be exportable resources, is if `resource[2]` is a dictionary.

File: awscli/customizations/cloudformation/artifact_exporter.py [hunk 3]
```diff
@@ -659,9 +659,23 @@ def export(self):
...
 
-        for resource_id, resource in self.template_dict["Resources"].items():
+        self.export_resources(self.template_dict["Resources"])
+
+        return self.template_dict
+
+    def export_resources(self, resource_dict):
+        for resource_id, resource in resource_dict.items():
+
+            if resource_id.startswith("Fn::ForEach"):
```
[the remark this answers is not in the document]
Comment (arthurboghossian) [reply awscli/customizations/cloudformation/artifact_exporter.py]: I've updated the `startswith` reference accordingly

File: awscli/customizations/cloudformation/artifact_exporter.py [hunk 4]
```diff
@@ -659,9 +659,23 @@ def export(self):
...
+                if not isinstance(resource, list) or len(resource) < 3:
+                    raise ValueError("Fn::ForEach with name {0} has invalid format".format(resource_id))
+                if isinstance(resource[2], dict):
+                    self.export_resources(resource[2])
+                continue
 
             resource_type = resource.get("Type", None)
+            if resource_type is None:
```
[the remark this answers is not in the document]
Comment (arthurboghossian) [reply awscli/customizations/cloudformation/artifact_exporter.py]: I had added this (removed it from the latest commit) to prevent the need to go through the items in `self.resources_to_export` if the `resource_type` resolves to None (see loop below)

```python
            resource_type = resource.get("Type", None)
            # ...
            for exporter_class in self.resources_to_export:
                if exporter_class.RESOURCE_TYPE != resource_type:
                    continue
                # ...
```

However, looking at the `Resource` class, if `RESOURCE_TYPE` is not defined for a subclass of a Resource, it would also resolve to `None`, thought I don't think we'd every have such a subclass, but it is possible, so opting to not have this change as part of this PR

File: awscli/customizations/cloudformation/artifact_exporter.py [hunk 5]
```diff
@@ -659,9 +659,23 @@ def export(self):
...
+        self.export_resources(self.template_dict["Resources"])
+
+        return self.template_dict
+
+    def export_resources(self, resource_dict):
+        for resource_id, resource in resource_dict.items():
+
+            if resource_id.startswith("Fn::ForEach"):
+                if not isinstance(resource, list) or len(resource) < 3:
+                    raise ValueError("Fn::ForEach with name {0} has invalid format".format(resource_id))
```
[the remark this answers is not in the document]
Comment (arthurboghossian) [reply awscli/customizations/cloudformation/artifact_exporter.py]: I've added the corresponding exception and modified the exception message accordingly

File: tests/unit/customizations/cloudformation/test_artifact_exporter.py [hunk 6]
```diff
@@ -1016,6 +1016,151 @@ def test_template_export(self, yaml_parse_mock):
...
+            exported_template = template_exporter.export()
+            self.assertEqual(exported_template, template_dict)
+
+            open_mock.assert_called_once_with(
+                    make_abs_path(parent_dir, template_path), "r")
+
+            self.assertEqual(1, yaml_parse_mock.call_count)
+
+            resource_type1_class.assert_called_with(self.s3_uploader_mock)
+            resource_type1_instance.export.assert_called_with(
```
[the remark this answers is not in the document]
Comment (arthurboghossian) [reply tests/unit/customizations/cloudformation/test_artifact_exporter.py]: I've modified the test accordingly

File: awscli/customizations/cloudformation/artifact_exporter.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hssyoo) [reply awscli/customizations/cloudformation/artifact_exporter.py:672]: I think @kyleknap's point was that even if a user supplies an invalid type for `resource[2]`, they're still expecting _something_ to happen. If we just ignore it and move on, we're silently failing the user's expectation. In this case it'd be preferable to explicitly throw an exception. We shouldn't attempt to fix/ignore bad input, any fixes should be made by the user.

File: awscli/customizations/cloudformation/artifact_exporter.py [hunk 7]
```diff
@@ -659,7 +659,19 @@ def export(self):
...
-        for resource_id, resource in self.template_dict["Resources"].items():
+        self.export_resources(self.template_dict["Resources"])
+
+        return self.template_dict
+
+    def export_resources(self, resource_dict):
+        for resource_id, resource in resource_dict.items():
+
+            if resource_id.startswith("Fn::ForEach::"):
+                if not isinstance(resource, list) or len(resource) < 3:
```
Comment (hssyoo) [inline awscli/customizations/cloudformation/artifact_exporter.py:670]: Can the length of `resource` be greater than 3? My understanding is that the list contains an id, collection, and output dict.

File: awscli/customizations/cloudformation/artifact_exporter.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (arthurboghossian) [reply awscli/customizations/cloudformation/artifact_exporter.py]: That makes sense. I've modified the condition accordingly

File: awscli/customizations/cloudformation/artifact_exporter.py (hunk 7 above)
Comment (arthurboghossian) [reply awscli/customizations/cloudformation/artifact_exporter.py]: The length of `resource` currently can't be greater than 3, but there can be enhancements that could make it larger than 3 in the future, but for now, modified this condition to ensure the size of `resource` is exactly 3

**Review by hssyoo — APPROVED** [review]
🚀

**Comment by rhbecker** [comment]
A fix very similar to what this PR addressed is [needed](https://github.com/aws/aws-sam-cli/issues/5647) for the sam cli. Is there any chance that the authors of this PR would be willing to take that on?

**Comment by dcabib** [comment]
Hey everyone! 🎉

Just opened a PR that fixes this issue: [PR ##5647](https://github.com/aws/aws-sam-cli/pull/8298)

**TL;DR:** SAM CLI now supports Fn::ForEach! No more AttributeError crashes.

We followed the same approach AWS CLI took in their PR aws/aws-cli#8096 - detect ForEach constructs, skip them during local parsing, and let CloudFormation handle the expansion server-side (where it belongs).

Tested thoroughly:
- ✅ 5,870 unit tests passing
- ✅ 94% code coverage
- ✅ Real templates with Fn::ForEach validate and build successfully
- ✅ ForEach constructs preserved for CloudFormation

Been waiting 2.5 years for this one - should be good to go! 🚀
